### PR TITLE
[node-manager] Fail early-oom gracefully

### PR DESCRIPTION
**Pull request by zuzzas**
## Description

Fail with a helpful log message and generate an alert if PSI subsystem is unavailable.

## Why do we need it, and what problem does it solve?

There are OS versions that have PSI disabled by default.

## What is the expected result?

Pod is not in CrashloopBackOff and an alert is generated.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: node-manager
type: fix
summary: Fail early-oom gracefully with a helpful log message and a low-severity alert if PSI subsystem is unavailable.
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've started the workflow for label 'e2e/run/yandex-cloud'...

File: modules/040-node-manager/images/early-oom/src/psi-monitor.go [hunk 1]
```diff
@@ -35,16 +40,42 @@ const (
 	sysrqOOMCharacter = "f"
 )
 
+var psiAvailable = prometheus.NewGauge(prometheus.GaugeOpts{
```
Comment (yalosev) [inline modules/040-node-manager/images/early-oom/src/psi-monitor.go:43]: `psiUnavailable` ?

File: modules/040-node-manager/images/early-oom/src/psi-monitor.go [hunk 2]
```diff
@@ -55,6 +86,25 @@ func main() {
 	}
 }
 
+func probePSISupport() error {
+	fs, err := procfs.NewDefaultFS()
+	if err != nil {
+		log.Fatal(err)
```
Comment (yalosev) [inline modules/040-node-manager/images/early-oom/src/psi-monitor.go:92]: you have error handling but fail with log.Fatal. Is it necessarily?

File: modules/040-node-manager/images/early-oom/src/psi-monitor.go [hunk 3]
```diff
@@ -35,16 +40,42 @@ const (
...
+	go func() {
+		if err := server.ListenAndServe(); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	psiSupportErr := probePSISupport()
+	if psiSupportErr != nil {
+		log.Println(psiSupportErr)
+		psiAvailable.Set(1)
```
Comment (yalosev) [inline modules/040-node-manager/images/early-oom/src/psi-monitor.go:68]: Because this one is reading in an opposite meaning

Comment (zuzzas) [reply modules/040-node-manager/images/early-oom/src/psi-monitor.go]: Fixed.

File: modules/040-node-manager/images/early-oom/src/psi-monitor.go (hunk 2 above)
Comment (zuzzas) [reply modules/040-node-manager/images/early-oom/src/psi-monitor.go]: Not necessary. Bad copypasta while refactoring.

File: modules/040-node-manager/images/early-oom/src/psi-monitor.go (hunk 1 above)
Comment (zuzzas) [reply modules/040-node-manager/images/early-oom/src/psi-monitor.go]: Fixed.

**Review by yalosev — APPROVED** [review]

File: modules/040-node-manager/images/early-oom/src/psi-monitor.go [hunk 4]
```diff
@@ -35,16 +40,42 @@ const (
...
+		}
+	}()
+
+	psiSupportErr := probePSISupport()
+	if psiSupportErr != nil {
+		log.Println(psiSupportErr)
+		psiUnavailable.Set(1)
+
+		sig := <-c
+		shutdown(sig, server)
```
Comment (RomanenkoDenys) [inline modules/040-node-manager/images/early-oom/src/psi-monitor.go:71]: One question. When psiSupportErr is not nil, do we set metric and wait SIGTERM from kubelet due to liveness probe fails ?  Is this expected behavior ?

Comment (RomanenkoDenys) [reply modules/040-node-manager/images/early-oom/src/psi-monitor.go:71]: Or we should set metric and do nothing in main cycle ? If our pod is never be ready we will get unwanted alerts about it (KubernetesDaemonSetReplicasUnavailable, KubernetesDaemonSetNotUpToDate).

**Review by RomanenkoDenys — COMMENTED** [review]
added one little question

File: modules/040-node-manager/images/early-oom/src/psi-monitor.go (hunk 4 above)
Comment (yalosev) [reply modules/040-node-manager/images/early-oom/src/psi-monitor.go:71]: I guess we can remove `sig := <-c` and everything would be fine

Comment (yalosev) [reply modules/040-node-manager/images/early-oom/src/psi-monitor.go:71]: Or maybe we should check PSISupport before starting the http server?

Comment (RomanenkoDenys) [reply modules/040-node-manager/images/early-oom/src/psi-monitor.go:71]: No, all is ok. Liveness probe checked from kube-rbac-proxy so this code works as expected. If we remove 70 line, we go to the cycle which tries to do oom work.

**Review by RomanenkoDenys — APPROVED** [review]
LGTM

**Comment by deckhouse-BOaTswain** [comment]
Cherry pick PR [2516](https://github.com/deckhouse/deckhouse/pull/2516) to the branch [release-1.36](https://github.com/deckhouse/deckhouse/tree/release-1.36) successful!

**Pull request merged**